proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,337 @@
+with enough rent-exempt lamports to initialize state. The processor will
+allocate, assign, and initialize the account if it does not exist.
+
+The `StageFeatureForActivation` instruction is structured as follows:
+
+- Data: None
+- Accounts:
+  - Feature account
+  - Staged Features PDA: writable
+  - Multi-signature authority: signer
+  - System program (required only for initializing)
+
+Note that features can only be staged in the epoch prior to the target
+activation epoch. This means a feature staged by invoking
+`StageFeatureForActivation` in epoch `N-1` is scheduled for activation (through
+the process below) at the end of epoch `N`. This is checked by the `Clock`
+sysvar.
+
+When a feature is staged for activation, initial stake support is zero.
+
+#### Staged Features PDA State
+
+A Staged Features PDA will be created for each epoch in which features are
+staged to be activated. If no features are staged for a given epoch, that
+epoch's corresponding Staged Features PDA will never be initialized and thus
+will not exist.
+
+These PDAs will not be garbage collected and can be referenced for historical
+purposes.
+
+The address of the Staged Features PDA for a given epoch is derived as follows,
+where `epoch` is a `u64` serialized to eight little-endian bytes:
+
+```
+"staged_features" + <epoch>
+```
+
+The data for the Staged Features PDA will be structured as follows:
+
+```c
+#define FEATURE_ID_SIZE 32
+#define MAX_FEATURES 8
+
+/**
+ * A Feature ID and its corresponding stake support, as signalled by validators.
+ */
+typedef struct {
+    /** Feature identifier (32 bytes for public key). */
+    uint8_t feature_id[FEATURE_ID_SIZE];
+    /** Stake support (u64 serialized to little-endian). */
+    uint8_t stake_support[8];
+} FeatureStake;
+
+/**
+ * Staged features for activation. 
+ */
+typedef struct {
+    /**
+     * Features staged for activation at the end of the current epoch, with
+     * their corresponding signalled stake support.
+     */
+    FeatureStake features[MAX_FEATURES];
+} StagedFeatures;
+```
+
+As depicted in the above layout, a maximum of 8 features can be staged for a
+given epoch.
+
+### Step 3: Signaling Support for Staged Features
+
+With an on-chain reference point to determine the features staged for activation
+for a particular epoch, nodes will signal their support for the staged features
+supported by their software.
+
+A node signals its support for staged features by invoking another new Feature
+Gate program instruction: `SignalSupportForStagedFeatures`. This instruction
+expects the Validator Support Signal PDA (defined below) to either exist or be
+funded with enough rent-exempt lamports to initialize state. The processor will
+allocate, assign, and initialize the account if it does not exist.
+
+The `SignalSupportForStagedFeatures` instruction is structured as follows:
+
+- Data: A `u8` bit mask of the staged features.
```

Comment:
Does this also need to include the epoch? As the list of staged features changes epoch-to-epoch, we need a way of ensuring that the bit mask in each validator support signal transaction is interpreted for the correct epoch. If the transaction is delayed or is executed in a different epoch than the one it was sent, for whatever reason, then the support signal will be mis-interpreted.

Reply:
I would guess the `Staged Features PDA` supplied should indicate the intended epoch?

Reply:
It's in the PDA for the validator's support signal. When they send a bitmask, it gets recorded in their signal PDA, and if they send again, it's overwritten. The current epoch is checked via `Clock` sysvar when the program is invoked with this instruction.